common/lib/core-interfaces/src/provider.ts

```diff
@@ -0,0 +1,70 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * @internal
+ * This utility type is meant for internal use by @see FluidObject
+ * Produces a valid FluidObject key given a type and a property.
+ * A valid FluidObject key is a property that exists on the incoming type
+ * as well as on the type of the property itself. For example, IProvideFoo.IFoo.IFoo
+ * This aligns with the FluidObject pattern expected to be used with all FluidObjects.
+ * For example:
+ * ```
+ * interface IProvideFoo{
+ *  IFoo: IFoo
+ * }
+ * interface IFoo extends IProvideFoo{
+ *  foobar();
+ * }
+ * ```
+ * This pattern enables discovery, and delegation in a standard way which is central
```

Comment:
Really glad you found a good a way to accommodate proper delegation, yay!  Can you update this doc comment to cover at least briefly about using `Partial<IProvideFoo>` if you want to do proper delegation?  And I'm with Vlad - mentioning `IProvideFoo.IFoo.IFoo` doesn't necessarily bring clarity.  That is indeed how the type magic works (it finds any prop that also exists on the prop's type), but not how this stuff is ever used.